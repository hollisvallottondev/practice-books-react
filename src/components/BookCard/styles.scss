@import '../../colors.scss'; 

.cardContainer {
	display: flex;
	flex-direction: column;
	justify-content: space-between;
	background-color: $secondaryColor;
	color: $primaryColor;
	font-size: calc(0.75vh + 0.75vw);
	margin-bottom: 15px;
	transition: border 0.2s linear;

	@extend .container;
	
	.posterContainer {
		width: 200px;
	}
	.detailsContainer {
		width: 100%;
		.cardHeader {
			display: flex;
			flex-direction: row;
			justify-content: space-between;

			h3 {
				color: $accentColor;
				margin: 0;
				margin-bottom: 20px;
			}

			.rating {
				justify-self: flex-start;
				display: flex;
				justify-content: center;
				align-items: center;
				min-width: 35px;
				min-height: 20px;
				max-width: 35px;
				max-height: 20px;
				padding: 5px;
				color: $secondaryColor;
				background-color: $accentColor;
				border-radius: 5px;
			}
		}

		.cardDescription {
			display: flex;
			flex-direction: column;
			justify-content: space-evenly;
			div {
				display: flex;
				justify-content: flex-start;
				p {
					text-align: justify;
					margin: 0;
					margin-bottom: 20px;
					flex: 2;
				}
	
				label {
					font-weight: bold;
					margin-bottom: 5px;
					cursor: inherit;
					flex: 1;
				}

				input {
					flex: 1;
					margin-bottom: 20px;
				}
			}

		}
	}


}

.hover {
	&:hover{
		border: 5px solid $accentColor;
		cursor: pointer;
		// background-color: $primaryColor;
		color: $accentColor;
	}
}

.container {
	padding: 20px;
}

.deleteContainer {
	display: flex;
	flex-direction: column;
	align-content: center;
	justify-content: center;

	p {
		height: 30px;
		width: auto;
		text-align: center;
	}
}

@import '../../main.scss';