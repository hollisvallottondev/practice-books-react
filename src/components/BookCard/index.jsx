import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import styles from "./styles.scss";

const BookCard = ({
  handleBookSelected,
  handleCancel,
  handleBookSave,
  handleBookDelete,
  index,
  book,
  edit,
}) => {
  const [bookValues, setBookValues] = useState({});

  const handleSelect = (index) => {
    if (handleBookSelected) handleBookSelected(index);
  };
  const handleChange = (key, value) => {
    setBookValues({ ...bookValues, [key]: value });
  };

  const handleDelete = (event) => {
    handleBookDelete(index);
    event.stopPropagation();
  };

  useEffect(() => {
    setBookValues(book);
  }, [book]);

  const { id, ...bookProps } = bookValues;

  return (
    <div
      className={`${styles.cardContainer} ${
        handleBookSelected ? styles.hover : ""
      }`}
      onClick={() => handleSelect(index)}
    >
      <div className={styles.detailsContainer}>
        <div className={styles.cardHeader}>
          <h3>{bookValues.title}</h3>
        </div>
        <div className={styles.cardDescription}>
          {Object.keys(bookProps).map((key) => {
            if ((key === "title" && edit) || key !== "title") {
              return (
                <div key={key}>
                  <label>
                    {key[0].toUpperCase() +
                      key.replace("_", " ").slice(1, key.length) +
                      ":"}
                  </label>
                  {edit ? (
                    <input
                      onChange={({ target }) => handleChange(key, target.value)}
                      value={bookProps[key]}
                    />
                  ) : (
                    <p>{bookProps[key] || `No ${key} available`}</p>
                  )}
                </div>
              );
            }
          })}
        </div>
      </div>
      {edit ? (
        <div className={styles.actions}>
          <button onClick={handleCancel} className={styles.negative}>
            Cancel
          </button>
          <button
            onClick={() => handleBookSave(bookValues, index)}
            className={styles.positive}
          >
            Save
          </button>
        </div>
      ) : (
        <div className={styles.actions}>
          <button onClick={handleDelete} className={styles.negative}>
            Delete
          </button>
        </div>
      )}
    </div>
  );
};

BookCard.propTypes = {
	handleBookSelected: PropTypes.func,
	handleCancel: PropTypes.func,
	handleBookSave: PropTypes.func,
	handleBookDelete: PropTypes.func,
	index: PropTypes.number,
	book: PropTypes.object.isRequired,
	edit: PropTypes.bool,
  };
  
export default BookCard;

export const ConfirmDelete = ({book, handleModalClose, handleConfirmDelete, focusDeleteIndex}) => (
  <div className={styles.deleteContainer}>
    {book && (
      <p>{`Are you sure you want to delete ${book.title} by ${book.author}?`}</p>
    )}
    <div className={styles.actions}>
      <button className={styles.negative} onClick={() => handleModalClose()}>
        Cancel
      </button>
      <button
        className={styles.positive}
        onClick={() => handleConfirmDelete(focusDeleteIndex)}
      >
        Confirm
      </button>
    </div>
  </div>
);

export const EditBook = ({book, handleCancel, handleBookSave}) => (
	<div>
		<BookCard 
			book = {book} 
			imgSize={"w200"} 
			index = {null} 
			edit={true} 
			handleCancel={handleCancel} 					
			handleBookSave={handleBookSave}
		/>
	</div>
);


