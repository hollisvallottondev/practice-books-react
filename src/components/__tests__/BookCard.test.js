import React from 'react';
import BookCard from '../BookCard/index.jsx';
import renderer from 'react-test-renderer';
import booksMocks from '../../mocks/books_seed.json';

test('BookCard tree renders correctly', () => {
	const index = 0;
	const book = booksMocks[index];
	const handleBookSelected = jest.fn(() => console.log('Handle book selected'));
	const handleBookSave = jest.fn(() => console.log('Handle book save')); 
	const handleBookDelete = jest.fn(() => console.log('Handle book delete')); 

	const component = renderer.create(
		<BookCard
			key={book.id} 
			handleBookSelected={handleBookSelected} 
			handleBookSave={handleBookSave}
			handleBookDelete={handleBookDelete}
	    	book={book} 
			imgSize={"w200"} 
			index={index} />
	  );
	
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});