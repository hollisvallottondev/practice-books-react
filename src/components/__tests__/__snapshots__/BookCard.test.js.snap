// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard tree renders correctly 1`] = `
<div
  className="undefined undefined"
  onClick={[Function]}
>
  <div>
    <div>
      <h3 />
    </div>
    <div />
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
